dotnet/src/Connectors/Connectors.Google/Core/Gemini/Clients/GeminiChatCompletionClient.cs

```diff
@@ -226,15 +226,33 @@ internal sealed class GeminiChatCompletionClient : ClientBase
 
         for (state.Iteration = 1; ; state.Iteration++)
         {
-            using var httpRequestMessage = await this.CreateHttpRequestAsync(state.GeminiRequest, this._chatStreamingEndpoint).ConfigureAwait(false);
-            using var response = await this.SendRequestAndGetResponseImmediatelyAfterHeadersReadAsync(httpRequestMessage, cancellationToken)
-                .ConfigureAwait(false);
-            using var responseStream = await response.Content.ReadAsStreamAndTranslateExceptionAsync()
-                .ConfigureAwait(false);
-
-            await foreach (var messageContent in this.GetStreamingChatMessageContentsOrPopulateStateForToolCallingAsync(state, responseStream, cancellationToken).ConfigureAwait(false))
+            using (var activity = ModelDiagnostics.StartCompletionActivity(
+                this._chatGenerationEndpoint, this._modelId, ModelProvider, chatHistory, executionSettings))
             {
-                yield return messageContent;
+                HttpResponseMessage httpResponseMessage;
+                Stream responseStream;
+                try
+                {
+                    using var httpRequestMessage = await this.CreateHttpRequestAsync(state.GeminiRequest, this._chatStreamingEndpoint).ConfigureAwait(false);
+                    // We cannot dispose these two objects leaving the try-catch block because we need them to read the response stream
+                    httpResponseMessage = await this.SendRequestAndGetResponseImmediatelyAfterHeadersReadAsync(httpRequestMessage, cancellationToken).ConfigureAwait(false);
+                    responseStream = await httpResponseMessage.Content.ReadAsStreamAndTranslateExceptionAsync().ConfigureAwait(false);
+                }
+                catch (Exception ex)
+                {
+                    activity?.SetError(ex);
+                    throw;
+                }
```

Comment:
What about exceptions that might emerge from the foreach loop below?

Reply:
I will dispose them in the catch block.

Reply:
I think we can only rely on exception handling further up the stack in this case, which is outside the scope of these activities.

Reply:
I don't understand why. Can you elaborate?

Reply:
Since we cannot yield return in a try-catch block, we can't catch exceptions raised within the foreach loop and associate them to the activities.

Reply:
You can, you just can't use the nice foreach syntax. Instead of something like:
```C#
await foreach (var x in Foo())
    yield return x;
```
you'd write:
```C#
IAsyncEnumerator<T> e = Foo().GetEnumerator();
try
{
while (true)
{
    try
    {
        if (!await e.MoveNextAsync()) break;
    }
    catch (Exception exc) { ... }

    yield return e.Current;
}
}
finally
{
    await e.DisposeAsync();
}
```
We can decide it's not worth it, but it is possible.

Reply:
Thanks! I think this is ok. Should be straightforward. I will make the changes.